accepted/00000-modular-repos-with-clm.md

```diff
@@ -0,0 +1,180 @@
+> populated by the module build system.
+
+*modulemd specification*
+
+This situation is handled with the following additional procedure:
+
+Instead of a plain list of RPMs, the API returns a map of package names provided in the `api` section to non-empty lists of artifacts that match the corresponding package names. If a module lists a name in the `api` section but does not provide any artifacts, that name will not be included in the map.
+
+Since not all of the artifacts are listed in the `api` section, the remaining artifacts are put into a special "null" key.
+
+For each key of the map, the backend gets the packages with the same name from all available sources. If any of the packages are not included in the corresponding artifacts list, the package is added to the blacklist.
+
+As a result, when a module defines a package name in its API *and* provides an actual RPM with it, all the other RPMs with the same name are effectively blacklisted even if they are in a different source.
+
+### Module updates and multiple versions
+Every released update of a module is added into the module metadata as a separate module entry with a new `version` number. By design, Uyuni needs to include either all versions of a module or none at all.
+This can be achieved by enabling all versions of the module in the API at the same time. As a result, all the artifacts from all versions will be included sharing the same key in the returning map.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+For each build of the project, the backend needs a full list of RPMs that each module provides. The most important drawback is the amount of data to process for each build considering the module metadata will grow constantly with each released update.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The current version of Uyuni supports modular repositories in the most basic way. Modular repositories are recognized, mirrored correctly and can be assigned systems. However, doing package operations on these systems through the Uyuni UI is not supported. These systems must be managed from their own package managers.
+
+The next step to supporting modular repositories would be to implement modularity in the core with dedicated logic for package operations and a new set of UI. The most obvious benefit of this would be to be able to do package operations through the UI while still providing modular repositories to the clients. However, this implementation would require much more effort and currently beyond the scope.
+
+# Further enhancements
+[enhancements]: #enhancements
+
+### Interactive inputs for filters
+To provide better user experience, the filter inputs can be turned into responsive dropdowns showing a list of module/stream names with realtime feedback.
+However, since the module list can only be retrieved from the project sources and the current implementation of CLM filters are independent of the sources, this is not currently achievable.
```

Comment:
What would we need to change in our current implementation of CLM filters? Would it also enhance the experience for other OSs?

Reply:
We will add a new type of filter and the additional logic to process this filter. The current experience won't be affected in any way.

Reply:
We would need to relate filters to projects so that a filter cannot exist outside of a project's scope. But I believe that's a radical change in the current CLM implementation.

Reply:
Well, but I think it makes sense to think about it. Here we have the first real use case where it makes sense to create a filter for a specific project.
But on the other hand this is not really true. It is more a filter for specific source selection.

The idea could be to add a validator to every filter. Such a filter can only be added to a project if the validation succeed. Otherwise we could not display it in the list of filters in a project.

Reply:
> The idea could be to add a validator to every filter. Such a filter can only be added to a project if the validation succeed. Otherwise we could not display it in the list of filters in a project.

This is a more realistic approach IMO. We can validate the project setup at every modification and notify the user in real-time.

Reply:
As long as we can keep the filter detached from the project context, I'd keep it. We started thinking about it and early feedback were "no please, generic filters to be reused are the most important one". Then we got the request for filtering at an Environment level. As always there is at least a use case to have a different implementation, but I'd keep the feature *as simple as possible*, and it is already getting complexity every day.
So to me the solution is to keep the filter context separated from the project: as for the ActivationKey workflow, we can populate the first dropdown box with all possible modular repositories. Based on this first selection, we present a second filter step, where the user can choose the modules he wants. Here at this point we can do the validity check, and we can even disable saving the filter if a bad selection is attempted. Or we flag the filter as *in error* and we reflect this information in projects where the filter is attached, with the status messages mentioned in the `### Error reporting & feedback interface` section of this RFC.